File: WOT-iOS/WOT-iOS/core/datasources/WOTTree/WOTTree+Tanks/WOTNode+ModuleTree.m
//
//  WOTNode+Tanks.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 7/15/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTNode+ModuleTree.h"
#import "ModulesTree+UI.h"
#import "ModulesTree+FillProperties.h"
#import <objc/runtime.h>

@implementation WOTNode (ModuleTree)


- (id)initWithModuleTree:(ModulesTree *)module {

    NSURL *imageURL = [module localImageURL];
    WOTNode *result = [self initWithName:module.name imageURL:imageURL];
    if (result){
        
        [result setModuleTree:module];
    }
    return result;
}

- (WOTModuleType) moduleType {

    return [self moduleTree].moduleType;
}

static const void *WOTModuleTree = &WOTModuleTree;
- (void)setModuleTree:(ModulesTree *)moduleTree {
    
    objc_setAssociatedObject(self, WOTModuleTree, moduleTree, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (ModulesTree *)moduleTree {

    return objc_getAssociatedObject(self, WOTModuleTree);
}

@end

File: WOT-iOS/WOT-iOS/core/datasources/WOTTree/WOTTree+Tanks/WOTNode+ModuleTree.swift
//
//  WOTNode+ModuleTree.swift
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 7/17/18.
//  Copyright © 2018 Pavel Yeshchyk. All rights reserved.
//

import Foundation

protocol WOTNodeModuleTreeProtocol {
    var moduleTree: ModulesTree? { get set }
    var moduleType: WOTModuleType? { get }
}

struct AssociationKeys {
    static var moduleTreeKey: Int = 0
}

extension WOTNodeSwift: WOTNodeModuleTreeProtocol {
    var moduleTree: ModulesTree? {
        get {

            return objc_getAssociatedObject(self, &AssociationKeys.moduleTreeKey) as? ModulesTree

        }

        set {

            objc_setAssociatedObject(self, &AssociationKeys.moduleTreeKey, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }

    }
    var moduleType: WOTModuleType? {
        return self.moduleTree?.moduleType()

    }

}

//@property (nonatomic, strong)ModulesTree *moduleTree;
//@property (nonatomic, readonly)WOTModuleType moduleType;
//
//- (id)initWithModuleTree:(ModulesTree *)module;
//- (void)setModuleTree:(ModulesTree *)moduleTree;
//- (ModulesTree *)moduleTree;
